### PR TITLE
New course: 20937

### PR DESCRIPTION
**Pull request by jonohayon**
Adds the course "Defensive systems programming" (ID 20937) to the repository, along with a complete implementation of maman 11.

### CODE REVIEW

File: 20937/maman-11/USocial/User.cpp [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        }
+    }
+}
+
+void User::receiveMessage(const Message *message) {
+    if (nullptr == message) {
+        throw UserException("Tried to receive a null pointer message!");
+    }
+
+    this->receivedMessages.emplace_back(std::move(message));
```
Comment (Magal-W) [inline 20937/maman-11/USocial/User.cpp:106]: You don't need to move pointers

File: 20937/maman-11/USocial/User.cpp [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        std::begin(this->friends),
+        std::end(this->friends),
+        user->getId()
+    );
+
+    if (listUser == std::end(this->friends)) {
+        throw UserException("Regular users are allowed to send messages only to their friends!");
+    }
+
+    user->receiveMessage(message);
```
Comment (Magal-W) [inline 20937/maman-11/USocial/User.cpp:132]: This is dangerous. If a user tries to send a message to someone who isn't their friend, the message is leaked

File: 20937/maman-11/USocial/Post.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    const Media *media;
+
+public:
+    /**
+     * Constructs a new post.
+     * @param text The text content of the post.
+     * @param media A pointer to an already allocated media piece for the post. The ownership of this parameter is
+     * passed to the newly created post.
+     */
+    Post(const std::string &text, const Media *media = nullptr) : text(text), media(std::move(media)) {};
```
Comment (Magal-W) [inline 20937/maman-11/USocial/Post.h:26]: Constructing in h file is weird if it's a big class and there are other methods
+ You still don't need to move pointers

Comment (jonohayon) [reply 20937/maman-11/USocial/Post.h:26]: Fixed

File: 20937/maman-11/USocial/User.cpp (hunk 1 above)
Comment (jonohayon) [reply 20937/maman-11/USocial/User.cpp:106]: Fixed

File: 20937/maman-11/USocial/User.cpp (hunk 2 above)
Comment (jonohayon) [reply 20937/maman-11/USocial/User.cpp:132]: Technically, you would be correct. But it is defined in the assignment that if user A sends a message to user B, user B is the owner of the message object and is in charge of freeing it.

I added a clause before the throw for deleting the message.

**Pull request merged**